main.go

```
@@ -41,42 +43,111 @@ func (v *Version) String() string {
+var appVersion = fmt.Sprintf("%s version: %s", core.AppName(), Ver.String())
 
-	config, args, err := loadConfig(appName)
+func main() {
+	config, parser, err := loadConfig()
 	if err != nil {
+		handleParseError(err, parser)
 		os.Exit(1)
 	}
+	if config == nil {
+		os.Exit(0)
+	}
 
-	if config.HTTPMode {
-		fmt.Println("Running in http mode")
-		enterHttpMode(config)
-	} else {
-		enterCliMode(appName, config, args)
+	// Show version and exit if the version flag was specified
+	if config.ShowVersion {
+		fmt.Println(appVersion)
+		os.Exit(0)
 	}
-}
 
-func enterHttpMode(config *config) {
 	client, err := walletrpcclient.New(config.WalletRPCServer, config.RPCCert, config.NoDaemonTLS)
 	if err != nil {
 		fmt.Fprintln(os.Stderr, "Error connecting to RPC server")
 		fmt.Fprintln(os.Stderr, err.Error())
 		os.Exit(1)
 	}
 
+	if config.HTTPMode {
+		fmt.Println("Running in http mode")
+		enterHTTPMode(config, client)
+	} else {
+		enterCliMode(config, client)
+	}
+}
+
+func enterHTTPMode(config *core.Config, client *walletrpcclient.Client) {
 	web.StartHttpServer(config.HTTPServerAddress, client)
 }
 
-func enterCliMode(appName string, config *config, args []string) {
-	client, err := walletrpcclient.New(config.WalletRPCServer, config.RPCCert, config.NoDaemonTLS)
-	if err != nil {
-		fmt.Fprintln(os.Stderr, "Error connecting to RPC server")
-		fmt.Fprintln(os.Stderr, err.Error())
+func enterCliMode(config *core.Config, client *walletrpcclient.Client) {
+	cli.Setup(client)
+	parser := flags.NewParser(&cli.DcrcliCommands, flags.Default)
+	if _, err := parser.Parse(); err != nil {
+		handleParseError(err, parser)
 		os.Exit(1)
 	}
+}
+
+func loadConfig() (*core.Config, *flags.Parser, error) {
+	// load defaults first
+	cfg := core.DefaultConfig()
+	commands := cli.DcrcliCommands
+	commands.Config = cfg
 
-	c := cli.New(client, appName)
-	c.RunCommand(args)
+	parser := flags.NewParser(&commands, flags.HelpFlag)
```

This is potentially problematic. Regardless of whether the user wants to run in web or cli mode, this code parses command-line options assuming we're running on cli. The following scenario highlights how this could be problematic:

Say the cli balance command uses a flag (`-e/--extended` | compare #62), and a user runs the following:
Scenario 1: `dcrcli send -e` -> error, unknown flag `e`
Scenario 2: `dcrcli balance -e` -> valid, `e` is a valid flag for the `balance` **cli** command
Scenario 3: `dcrcli balance -e --http` -> valid, because `e` is a valid flag for the `balance` command but this should ideally fail because `e` is an invalid flag for any command in http mode. http mode does not even support cli commands by the way.

The ideal behaviour for the Scenario 3 would have been to return an error that the flag `e` is unsupported in http mode. If this parser was created with just `&cfg` and not `&commands`, this flag would have been rejected

---

Parsing twice is one work-around. First parse is as basic as possible, does not include cli commands and command-specific options. This parse too would not throw error if it encounters an invalid option because such option may actually be a valid command-specific option if we're running in cli.
Compare https://github.com/itswisdomagain/dcrcli/blob/pr71-update/main.go#L14

First parse helps us know what mode we're running on, so second parse can be more thorough, including commands and command-specific options if running in cli and the reverse if not.
Compare https://github.com/itswisdomagain/dcrcli/blob/pr71-update/main.go#L39 and https://github.com/itswisdomagain/dcrcli/blob/pr71-update/main.go#L50

---

`--http` is a switch. Because it is also a top-level option, we encounter it immediately after the first parse. So, the program switches to web mode if the `--http` flag is encountered; every other command is ignored.
The current order of priority is as follows.
- -v displays the version information and exits. Every other flag or command is ignored.
- -h/--help
- --http